crates/ir-generated/src/emit/emitter.rs

```diff
@@ -0,0 +1,87 @@
+use std::mem;
+
+use rokugo_ir::{
+    container::IrContainer,
+    op_code::IrOpCode,
+    register::{chill::RegisterChill, natural::RegisterNat32, Register, RegisterId},
+};
+
+use crate::errors::register::RegisterAllocationError;
+
+use super::register_allocator::{RegisterAllocator, RegisterDropGuard};
+
+pub struct IrEmitter {
+    data: Vec<u8>,
+    register_allocator: RegisterAllocator,
+}
+
+impl IrEmitter {
+    pub fn new() -> Self {
+        IrEmitter {
+            data: Vec::new(),
+            register_allocator: RegisterAllocator::new(),
+        }
+    }
+}
+
+/// # Local Memory
+impl IrEmitter {
+    /// Allocates a virtual register, or prepare a native register to store a new 32-bit natural value.
+    pub fn alloc_register_nat32(
+        &mut self,
+        chill: RegisterChill,
+    ) -> Result<RegisterDropGuard<RegisterNat32>, RegisterAllocationError> {
+        let id = self.register_allocator.next_nat32()?;
+        self.emit(IrOpCode::AllocRegisterNat32);
+        self.emit_register_id(id);
+        self.data.extend_from_slice(&chill.to_le_bytes());
+
+        // SAFETY: This is safe, because `RegisterNat32` is a wrapper around `RegisterId`.
+        Ok(RegisterDropGuard::new(unsafe { mem::transmute(id) }))
```

Comment:
I believe this should instead be done through a constructor called `RegisterNat32::new_unchecked` or something. Relying on memory layout for things like this can be really leaky if we ever want to change the layout.

Reply:
But this function will be shared with additional crates via `pub` modifier. And can be used improperly, invalid using of API. I guess Rust do not have things like [internals visible to](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute?view=net-8.0).

Reply:
Yeah, that's what `unsafe` is for. It will still be better than using `std::mem::transmute` because it will not assume memory layouts to match.

Remember, *ensuring* is always preferred to *assuming*, and in this case we can pretty easily *ensure* the uses of the API are valid through the type system.

Reply:
API should be clear and simple, things like that can affect semver etc. This only confusing and I did not see necessary for make things like that public.

Reply:
By using `std::mem::transmute`, you are effectively making the ***memory layout*** of a struct affect semver. Especially given that Rust's memory layout is **not stable**, this smells to me like a horrible idea that's deemed to break.

Does it really hurt that much to add two stupidly simple functions to `RegisterNat32`'s API, in exchange for a lack of undefined behavior? Yeah, it's boilerplate, but who cares. You write it once and it's done.

Reply:
This only affects "semver" between these two crates, it does not related to public semver. What I said eariel, Rust do not have tools for that, or maybe I am do not know about them.

Reply:
We consulted that offline, and @liquidev said about make whole crate internal, what bring this tools for me for creating these functions. So I close this thread.